frontend/src/api/actions/init.js

```diff
@@ -32,140 +42,167 @@ function toggleMockApi(enableMockAPI) {
- * @returns
+ * @param {*} globals
+ * @returns {object} same as input but with `null` values removed.
  */
-function updateGlobals(enableMockAPI, newActiveSession, getState) {
-    const state = getState();
-    let globals = globalsSelector(state);
-
-    if (enableMockAPI != null) {
-        globals = {
-            ...globals,
-            mockAPI: enableMockAPI
-        };
-    }
-
-    if (newActiveSession) {
-        globals = {
-            ...globals,
-            activeSession: newActiveSession.id
-        };
+function prepareGlobals(globals) {
+    const ret = {};
+    for (const [key, val] of Object.entries(globals)) {
+        if (val != null) {
+            ret[key] = val;
+        }
     }
-
-    return setGlobals(globals);
+    return ret;
 }
 
 /**
- * The old way this is used to work is that after component mounted
- * actions that needs to be made asynchronously are executed in
- * `componentDidMount` of `App.js`, but leaving the `toggleMockAPI`
- * controlled by the mockAPI component.
- * However, this leads to a racing condition such that when mockAPI
- * should be turned on, the toggling action does not occur before all
- * the API calls are made. Hence the app is trying to fetch from the real
- * API instead of the mock API.
- * In addition, on the event of turning on and off the mock API, switching
- * active user role and setting active sessions, a similar consecutive API
- * actions are made comparing to the initialization.
+ * Various actions have side effects, requiring additional actions to
+ * be dispatched. For example, if a session changes, all data related to
+ * that session needs to be re-fetched.
  *
- * The new approach is to make each actions in `iniOrder` array, which
- * contains all the consecutive API actions that need to be made on
- * application load, as a stage, and move the actual toggling of the mock
- * API to this level instead of leaving it to the mockAPI component. Whenever
- * a related events occur (for example, changing active session), the action
- * creator will call this `initFromStage` action with the matching stage and
- * this function will perform different API actions that is necessary
- * after the specified `stage`
+ * This function allows you to specify "stage" to start the init procedure at.
+ * It will handle re-fetching any dependent data and dispatching any
+ * required actions depending on the stage specified.
  *
  * @export
- * @param {string} stage
- * @param {object} options
+ * @param {string} stage - What stage to start the init procedure at
+ * @param {{ startAfterStage: boolean }} options - if true, start from the stage following the specified stage; if false, start from the specified stage
  * @returns {function} an async function that handles all the API calls.
  */
-export function initFromStage(stage, options) {
+export function initFromStage(stage, options = { startAfterStage: false }) {
+    const startAfterStage = !!options.startAfterStage;
```

Comment:
Is there any occasion that `options.startAfterStage` is not a boolean?

Reply:
If someone passes in options but omits `startAfterStage`. If it is undefined, it will cause trouble when it is added.